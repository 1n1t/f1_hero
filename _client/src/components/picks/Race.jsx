import React, { Component } from 'react';
import { connect } from 'react-redux';
import axios from 'axios';

import Standings from './Standings';
import * as actions from './../../actions';

class Race extends Component {
  async componentDidMount() {
    this.setState({ raceWidth: this.refs.race.offsetWidth });
    const pick = (await axios.get(`/api/pick/${this.props.race.round}`)).data;

    if (pick) {
      this.setState({ pick });
    }
  }

  submitRace = pick => {
    this.setState({ pick });
  };

  fetchPick = () => {
    //	dispatch action and save number of selected race to the store
    this.props.selectRace(this.props.race.round);
  };

  displayPick() {
    const isPickSet =
      this.state && this.state.hasOwnProperty('pick') && this.state.pick;
    const isRaceSelected = this.props.selectedRace === this.props.race.round;

    if (isRaceSelected) {
      const data = {
        round: this.props.race.round,
        raceWidth: this.state.raceWidth,
        holderNode: this.props.holderNode
      };

      if (isPickSet) {
        return (
          <Standings
            list={this.state.pick.forecast}
            status={'submitted'}
            data={data}
          />
        );
      } else if (this.props.race.isPassed) {
        return <Standings status={'passed'} data={data} />;
      } else {
        return (
          <Standings
            list={this.props.drivers}
            status={'new'}
            data={data}
            onSubmit={this.submitRace}
          />
        );
      }
    }

    return '';
  }

  displayStatus() {
    if ((this.state && this.state.pick) || this.props.race.isPassed) {
      if (this.props.race.isPassed) {
        return <i className="fa fa-check" aria-hidden="true" />;
      } else {
        return <i className="fa fa-circle" aria-hidden="true" />;
      }
    } else {
      return <i className="fa fa-circle-o" aria-hidden="true" />;
    }
  }

  displayScore() {
    if (this.state && this.state.pick && this.props.race.isPassed) {
      return <p className="score">{`${this.state.pick.score}pt`}</p>;
    } else {
      return <p className="score">-</p>;
    }
  }

  render() {
    return (
      <div className="race-container" ref="race">
        <div className="race" onClick={this.fetchPick}>
          {this.displayStatus()}
          <div className="img-container">
            <img src={this.props.race.flagUrl} alt="" />
          </div>
          <p className="country-code">{this.props.race.alpha3code}</p>
          {this.displayScore()}
        </div>
        {this.displayPick()}
      </div>
    );
  }
}

function mapStateToProps({ selectedRace, races, drivers }) {
  return { selectedRace, races, drivers };
}

export default connect(mapStateToProps, actions)(Race);
