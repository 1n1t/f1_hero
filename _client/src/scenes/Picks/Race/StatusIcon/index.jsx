import React from 'react';
import PropTypes from 'prop-types';

const StatusIcon = ({ iconClass, status }) => (
	<i
		className={`fa ${iconClass}`}
		aria-hidden="true"
		data-test={`race-status-${status}`}
	/>
);

StatusIcon.propTypes = {
	iconClass: PropTypes.string.isRequired,
	status: PropTypes.oneOf(['new', 'submitted', 'passed']).isRequired
};

export default StatusIcon;
