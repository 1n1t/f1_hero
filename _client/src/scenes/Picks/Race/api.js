/* eslint-disable import/prefer-default-export */
import fetchApi from '../../../services/api';

const endPoints = {
	getPick(round) {
		return `/api/picks/${round}`;
	}
};

export const getPick = round => fetchApi(endPoints.getPick(round));
