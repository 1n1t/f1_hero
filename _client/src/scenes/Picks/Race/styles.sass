.Race
  text-align: center
  font-size: 14px
  cursor: pointer

  .fa
    margin-bottom: 10px

  .img-container
    width: 50px
    height: 33px

    img
      object-fit: cover
      object-position: 50% 50%
      width: 100%
      height: 100%

  .country-code
    margin: 10px 0

  .score
    margin: 10px 0 0 0
    font-weight: 300
