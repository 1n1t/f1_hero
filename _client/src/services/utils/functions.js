/* eslint-disable import/prefer-default-export */
export const isObjectEmpty = obj =>
	Object.keys(obj).length === 0 && obj.constructor === Object;

export const getNodePaddings = node => {
	const nodeStyle = window.getComputedStyle(node);
	return (
		parseInt(nodeStyle.getPropertyValue('padding-right'), 10) +
		parseInt(nodeStyle.getPropertyValue('padding-left'), 10)
	);
};
