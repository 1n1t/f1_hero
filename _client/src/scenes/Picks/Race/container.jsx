/* eslint-disable react/prop-types */
import React from 'react';
import { connect } from 'react-redux';

import Race from './index';
import { selectRace } from '../../../data/races/actions';

const RaceContainer = props => <Race {...props} />;

export const mapStateToProps = state => ({
	selectedRace: state.data.races.selected
});

export default connect(mapStateToProps, { selectRace })(RaceContainer);
