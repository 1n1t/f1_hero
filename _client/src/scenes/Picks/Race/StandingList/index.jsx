/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { arrayMove } from 'react-sortable-hoc';
import { calculatePickAndStemLeftPosition } from 'App/services/utils/functions';

import ForecastNew from './ForecastNew';
import ForecastMissed from './ForecastMissed';
import ForecastSubmitted from './ForecastSubmitted';
import SubmitButton from './SubmitButton';
import { postPick } from './api';
import { racePropType } from '../index';
import './styles.scss';

class StandingList extends Component {
	state = {
		pickPos: 0,
		stemPos: 0,
		list:
			this.props.status === 'submitted' ? this.props.list : this.props.drivers,
		status: this.props.status
	};

	componentDidMount() {
		const { races, round, raceNode, raceHolderNode } = this.props;
		const { pickPos, stemPos } = calculatePickAndStemLeftPosition(
			{
				picksNode: this.picks,
				stemNode: this.stem,
				raceNode,
				raceHolderNode
			},
			races.length,
			round
		);

		const newState = {
			...this.state,
			pickPos,
			stemPos
		};
		this.setState(newState);
	}

	onSortEnd = ({ oldIndex, newIndex }) => {
		const newList = arrayMove(this.state.list, oldIndex, newIndex);
		const formattedList = newList.map((value, index) => ({
			...value,
			position: index + 1
		}));

		const newState = {
			...this.state,
			list: formattedList //	todo make list array immutable
		};
		this.setState(newState);
	};

	submitPick = async () => {
		const { round, onSubmit, updateRace } = this.props;
		try {
			const { pick } = (await postPick(round, {
				round,
				forecast: this.state.list.slice(0, 10)
			})).data;

			const newState = {
				...this.state,
				status: 'submitted',
				list: pick.forecast
			};

			this.setState(newState);
			onSubmit(pick);
			updateRace({ round, field: 'hasPick', value: true });
		} catch (error) {
			//	todo display an error as a flash message
			console.error('error: unable to save pick!!', error);
		}
	};

	render() {
		const { status } = this.state;
		let Standings;

		switch (status) {
			case 'new':
				Standings = ForecastNew;
				break;
			case 'missed':
				Standings = ForecastMissed;
				break;
			default:
				Standings = ForecastSubmitted;
		}

		return (
			<div
				ref={el => {
					this.picks = el;
				}}
				className="StandingList"
				style={{ left: `${this.state.pickPos}px` }}
			>
				<div
					ref={el => {
						this.stem = el;
					}}
					className="stem"
					style={{ left: `${this.state.stemPos}px` }}
				/>
				<Standings
					items={this.state.list}
					onSortEnd={this.onSortEnd}
					lockToContainerEdges
					helperClass="draggable"
					lockAxis="y"
				/>
				{status === 'new' && <SubmitButton onClick={this.submitPick} />}
			</div>
			//	todo make function that will create refs
		);
	}
}

export const listItemPropTypes = {
	position: PropTypes.number.isRequired,
	_drivers: PropTypes.arrayOf(
		PropTypes.shape({
			code: PropTypes.string.isRequired,
			name: PropTypes.string.isRequired
		})
	)
};

StandingList.defaultProps = {
	list: [],
	drivers: [],
	onSubmit: () => {}
};

StandingList.propTypes = {
	status: PropTypes.oneOf(['new', 'submitted', 'passed']).isRequired,
	list: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	drivers: PropTypes.arrayOf(PropTypes.shape(listItemPropTypes)),
	round: PropTypes.number.isRequired,
	races: PropTypes.arrayOf(PropTypes.shape(racePropType)).isRequired,
	onSubmit: PropTypes.func,
	updateRace: PropTypes.func.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceHolderNode: PropTypes.object.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceNode: PropTypes.object.isRequired
};

export default StandingList;
