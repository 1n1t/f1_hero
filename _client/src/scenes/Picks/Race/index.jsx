/* eslint-disable jsx-a11y/click-events-have-key-events */
import React, { Component } from 'react';
import PropTypes from 'prop-types';

import { isObjectEmpty, makeRef } from 'App/services/utils/functions';
import { getPick } from './api';
import StandingList from './StandingList/container';
import StatusIcon from './StatusIcon';
import './styles.scss';

class Race extends Component {
	state = {
		pick: {}
	};

	setRacePick = pick => {
		const newState = {
			...this.state,
			pick: {
				...this.state.pick,
				...pick
			}
		};

		this.setState(newState);
	};

	fetchPick = async () => {
		const { round } = this.props.race;
		this.props.selectRace(round);

		const { pick } = (await getPick(round)).data;
		this.setRacePick(pick);
	};

	displayPick() {
		const { race, selectedRace, raceHolderNode } = this.props;
		const { pick } = this.state;
		const isPickSet = !isObjectEmpty(pick);
		const isRaceSelected = selectedRace.round === race.round;

		if (!isRaceSelected || (!isPickSet && race.hasPick)) {
			return null;
		}

		if (race.hasPick) {
			return (
				<StandingList
					list={pick.forecast}
					status="submitted"
					round={race.round}
					raceNode={this.raceNode}
					raceHolderNode={raceHolderNode}
				/>
			);
		} else if (race.isPassed) {
			return (
				<StandingList
					status="missed"
					round={race.round}
					raceNode={this.raceNode}
					raceHolderNode={raceHolderNode}
				/>
			);
		}
		return (
			<StandingList
				status="new"
				round={race.round}
				onSubmit={this.setRacePick}
				raceNode={this.raceNode}
				raceHolderNode={raceHolderNode}
			/>
		);
	}

	displayStatusIcon() {
		const { isPassed, hasPick } = this.props.race;

		// eslint-disable-next-line no-nested-ternary
		return isPassed ? (
			<StatusIcon iconClass="fa-check" status="passed" />
		) : hasPick ? (
			<StatusIcon iconClass="fa-circle" status="submitted" />
		) : (
			<StatusIcon iconClass="fa-circle-o" status="new" />
		);
	}

	render() {
		const { flagUrl, alpha3code, score } = this.props.race;
		const scoreText = score ? `${score}pt` : '-';

		return (
			<div
				ref={makeRef('raceNode', this)}
			>
				<div
					className="Race"
					onClick={this.fetchPick}
					role="link"
					tabIndex="0"
					data-test="race"
				>
					{this.displayStatusIcon()}
					<div className="img-container">
						<img src={flagUrl} alt={`flag of ${alpha3code}`} />
					</div>
					<p className="country-code">{alpha3code}</p>
					<p className="score">{scoreText}</p>
				</div>
				{this.displayPick()}
			</div>
		);
	}
}

export const racePropType = {
	round: PropTypes.number.isRequired,
	isPassed: PropTypes.bool.isRequired,
	hasPick: PropTypes.bool.isRequired,
	flagUrl: PropTypes.string.isRequired,
	alpha3code: PropTypes.string.isRequired,
	score: PropTypes.number
};

Race.propTypes = {
	race: PropTypes.shape(racePropType).isRequired,
	selectedRace: PropTypes.shape({
		round: PropTypes.number.isRequired
	}).isRequired,
	selectRace: PropTypes.func.isRequired,
	// eslint-disable-next-line react/forbid-prop-types
	raceHolderNode: PropTypes.object.isRequired
};

export default Race;
