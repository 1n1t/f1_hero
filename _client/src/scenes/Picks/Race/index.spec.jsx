import r from 'Root/jest/utils/renderer';
import racesSample from 'JestClient/samples/races';
import { sel } from 'JestClient/functions';

import Race from './index';

const defaultProps = {
	race: racesSample.all[0],
	selectedRace: racesSample.selected,
	selectRace() {}
};

r.init(Race, defaultProps);

it('renders correctly', () => {
	const wrapper = r.render();
	expect(wrapper).toMatchSnapshot();
});

describe('renders correct race status icon', () => {
	it('when race has passed', () => {
		const wrapper = r.render({
			race: { ...defaultProps.race, hasPick: false, isPassed: true }
		});
		const status = wrapper.find(sel('race-status-passed'));

		expect(status).toHaveLength(1);
	});

	it('when pick is submitted and race is not passed', () => {
		const wrapper = r.render({
			race: { ...defaultProps.race, hasPick: true, isPassed: false }
		});
		const status = wrapper.find(sel('race-status-submitted'));

		expect(status).toHaveLength(1);
	});

	it('when pick is not submitted and race is not passed', () => {
		const wrapper = r.render({
			race: { ...defaultProps.race, hasPick: false, isPassed: false }
		});
		const status = wrapper.find(sel('race-status-new'));

		expect(status).toHaveLength(1);
	});
});
