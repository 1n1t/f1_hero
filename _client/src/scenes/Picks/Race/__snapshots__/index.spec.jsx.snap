// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="Race"
    data-test="race"
    onClick={[Function]}
    role="link"
    tabIndex="0"
  >
    <StatusIcon
      iconClass="fa-check"
      status="passed"
    />
    <div
      className="img-container"
    >
      <img
        alt="flag of AUS"
        src="http://foo.bar"
      />
    </div>
    <p
      className="country-code"
    >
      AUS
    </p>
    <p
      className="score"
    >
      50pt
    </p>
  </div>
</div>
`;
